packages/flutter/lib/src/material/tabs.dart

```diff
@@ -760,543 +352,478 @@ class TabBar<T> extends Scrollable implements AppBarBottomWidget {
-        scrollTo(0.0);
+  void _initTabController() {
+    if (_controller != null)
+      _controller.animation.removeListener(_handleTick);
+    _controller = config.controller ?? DefaultTabController.of(context);
+    if (_controller != null) {
+      _controller.animation.addListener(_handleTick);
+      _changeAnimation = new  _ChangeAnimation(_controller);
+      _currentIndex = _controller.index;
+      final List<double> offsets = _indicatorPainter?.tabOffsets;
+      _indicatorPainter = new _IndicatorPainter(_controller)..tabOffsets = offsets;
     }
   }
 
   @override
-  void dispose() {
-    _selection?.removeAnimationListener(this);
-    super.dispose();
+  void dependenciesChanged() {
+    super.dependenciesChanged();
+    _initTabController();
   }
 
   @override
-  void handleSelectionDeactivate() {
-    _selection = null;
-  }
-
-  // Initialize _indicatorTween for interactive dragging between the tab on the left
-  // and the tab on the right. In this case _selection.animation.value is 0.5 when
-  // the indicator is below the selected tab, 0.0 when it's under the left tab, and 1.0
-  // when it's under the tab on the right.
-  void _initIndicatorTweenForDrag() {
-    assert(!_valueIsChanging);
-    int index = _selection.index;
-    int beginIndex = math.max(0, index - 1);
-    int endIndex = math.min(config.labels.length - 1, index + 1);
-    if (beginIndex == index || endIndex == index) {
-      _indicatorTween = new RectTween(
-        begin: _tabIndicatorRect(beginIndex),
-        end: _tabIndicatorRect(endIndex)
-      );
-    } else {
-      _indicatorTween = new _TabIndicatorTween(
-        begin: _tabIndicatorRect(beginIndex),
-        middle: _tabIndicatorRect(index),
-        end: _tabIndicatorRect(endIndex)
-      );
-    }
-  }
-
-  // Initialize _indicatorTween for animating the selected tab indicator from the
-  // previously selected tab to the newly selected one. In this case
-  // _selection.animation.value is 0.0 when the indicator is below the previously
-  // selected tab, and 1.0 when it's under the newly selected one.
-  void _initIndicatorTweenForAnimation() {
-    assert(_valueIsChanging);
-    _indicatorTween = new RectTween(
-      begin: _indicatorRect ?? _tabIndicatorRect(_selection.previousIndex),
-      end: _tabIndicatorRect(_selection.index)
-    );
+  void didUpdateConfig(TabBar oldConfig) {
+    super.didUpdateConfig(oldConfig);
+    if (config.controller != oldConfig.controller)
+      _initTabController();
   }
 
   @override
-  void handleStatusChange(AnimationStatus status) {
-    if (config.labels.isEmpty)
-      return;
-
-    if (_valueIsChanging && status == AnimationStatus.completed) {
-      _valueIsChanging = false;
-      setState(() {
-        _initIndicatorTweenForDrag();
-        _indicatorRect = _tabIndicatorRect(_selection.index);
-      });
-    }
+  void dispose() {
+    if (_controller != null)
+      _controller.animation.removeListener(_handleTick);
```

Comment:
I hesitated for a moment here because I was wondering whether you needed to dispose the controller.  I realize you don't want to do that because you don't "own" the controller, but it might be worth a comment for next time I read this code.